File: client/typescript/magnolia.test.ts
/**
 * Integration tests for MagnoliaClient and MagnoliaHMACClient.
 *
 * Run against a live server:
 *
 *   MAGNOLIA_BASE_URL=https://... MAGNOLIA_USERNAME=alice MAGNOLIA_PASSWORD=... \
 *       node --experimental-strip-types magnolia.test.ts
 *
 * Required env vars:
 *   MAGNOLIA_BASE_URL      server root, e.g. https://magnolia.example.com
 *   MAGNOLIA_USERNAME      login identifier (username or email)
 *   MAGNOLIA_PASSWORD      login password
 *
 * Optional env vars (related tests are skipped when absent):
 *   MAGNOLIA_PROXY_ID        )  both required for HMAC tests
 *   MAGNOLIA_HMAC_KEY        )
 *   MAGNOLIA_PROXY_USERNAME  )  both required for proxy session tests
 *   MAGNOLIA_PROXY_PASSWORD  )
 *   MAGNOLIA_MEDIA_FILE      path to any file; enables media upload/download tests
 *   MAGNOLIA_TARGET_USER_ID  enables conversation and message tests
 */

import { describe, it, before, after } from "node:test";
import assert from "node:assert/strict";
import { extname } from "node:path";
import { MagnoliaClient, MagnoliaHMACClient, MagnoliaProxySessionClient, APIError, type PostContentRequest } from "./magnolia.ts";

const TIMEOUT_MS = 10_000;
const PREFIX = `mgtest_${Math.floor(Date.now() / 1000)}`;

//  Helpers 

function env(key: string): string {
  return process.env[key] ?? "";
}

function skip(key: string): never {
  // node:test has no native skip-from-within; throw a recognisable sentinel.
  throw Object.assign(new Error(`SKIP: env var not set: ${key}`), { skip: true });
}

function requireEnv(...keys: string[]): void {
  for (const k of keys) if (!env(k)) skip(k);
}

function mediaTypeFor(filePath: string): string {
  const ext = extname(filePath).toLowerCase();
  if ([".jpg", ".jpeg", ".png", ".gif", ".webp"].includes(ext)) return "image";
  if ([".mp4", ".webm", ".mov"].includes(ext)) return "video";
  return "file";
}

/** Wraps a test body so that skip-sentinels are reported as skips, not failures. */
async function run(t: { skip: (msg?: string) => void }, fn: () => Promise<void>): Promise<void> {
  try {
    await fn();
  } catch (err: any) {
    if (err?.skip) {
      t.skip(err.message);
    } else {
      throw err;
    }
  }
}

//  State shared across tests 

let client: MagnoliaClient;
let userID = "";
let postID = "";
let commentID = "";
let mediaID = "";
let convID = "";
let msgID = "";

//  Session client tests 

describe("Session client", () => {
  before(async () => {
    requireEnv("MAGNOLIA_BASE_URL", "MAGNOLIA_USERNAME", "MAGNOLIA_PASSWORD");
    client = new MagnoliaClient(env("MAGNOLIA_BASE_URL"), { timeoutMs: TIMEOUT_MS });
    const resp = await client.login(env("MAGNOLIA_USERNAME"), env("MAGNOLIA_PASSWORD"));
    assert.ok(resp.user.user_id, "login returned empty user_id");
    userID = resp.user.user_id;
  });

  after(async () => {
    // Clean up in dependency order.
    if (msgID)     { try { await client.deleteMessage(msgID);       } catch {} }
    if (commentID) { try { await client.deleteComment(commentID);   } catch {} }
    if (mediaID)   { try { await client.deleteMedia(mediaID);       } catch {} }
    if (convID)    { try { await client.deleteConversation(convID); } catch {} }
    if (postID)    { try { await client.deletePost(postID);         } catch {} }
    try { await client.logout(); } catch {}
  });

  //  Auth 

  it("Auth/Me returns the logged-in user", async () => {
    const resp = await client.me();
    assert.equal(resp.user.user_id, userID);
  });

  //  Profile 

  it("Profile/Get returns the user's profile", async () => {
    const resp = await client.getProfile(userID);
    assert.equal(resp.user_id, userID);
  });

  it("Profile/Update round-trips bio and restores it", async () => {
    const orig = await client.getProfile(userID);
    const newBio = `${PREFIX}_bio`;
    try {
      const updated = await client.updateProfile({ bio: newBio });
      assert.equal(updated.bio, newBio);
    } finally {
      await client.updateProfile({ bio: orig.bio });
    }
  });

  //  Posts 

  it("Posts/Create saves a draft", async () => {
    const resp = await client.createPost(
      [{ content_type: "text", display_order: 0, content: `${PREFIX} post` }],
      false,
      [PREFIX],
    );
    assert.ok(resp.post_id, "empty post_id");
    assert.equal(resp.is_published, false);
    postID = resp.post_id;
  });

  it("Posts/Get retrieves the post", async () => {
    assert.ok(postID, "no postID - previous test skipped?");
    const resp = await client.getPost(postID);
    assert.equal(resp.post_id, postID);
  });

  it("Posts/Update changes content", async () => {
    assert.ok(postID);
    const newContent = `${PREFIX} updated`;
    const resp = await client.updatePost(postID, {
      contents: [{ content_type: "text", display_order: 0, content: newContent }],
    });
    assert.equal(resp.contents[0]!.content, newContent);
  });

  it("Posts/Publish toggles is_published", async () => {
    assert.ok(postID);
    const resp = await client.publishPost(postID);
    assert.equal(resp.is_published, true);
  });

  it("Posts/List returns an array", async () => {
    const resp = await client.listPosts({ limit: 5 });
    assert.ok(Array.isArray(resp.posts));
  });

  it("Posts/Search returns an array", async () => {
    const resp = await client.searchPosts({ q: PREFIX });
    assert.ok(Array.isArray(resp.posts));
  });

  //  Comments 

  it("Comments/Create posts a comment", async () => {
    assert.ok(postID);
    const resp = await client.createComment(postID, `${PREFIX} comment`);
    assert.ok(resp.comment_id, "empty comment_id");
    assert.equal(resp.content, `${PREFIX} comment`);
    commentID = resp.comment_id;
  });

  it("Comments/Update changes content", async () => {
    assert.ok(commentID);
    const edited = `${PREFIX} edited`;
    const resp = await client.updateComment(commentID, edited);
    assert.equal(resp.content, edited);
  });

  it("Comments/List includes the created comment", async () => {
    assert.ok(postID && commentID);
    const resp = await client.listComments(postID);
    assert.ok(resp.comments.some(c => c.comment_id === commentID));
  });

  //  Media 

  it("Media/Upload stores a file", async (t) => {
    await run(t, async () => {
      requireEnv("MAGNOLIA_MEDIA_FILE");
      const resp = await client.uploadMediaFile(env("MAGNOLIA_MEDIA_FILE"));
      assert.ok(resp.media_id, "empty media_id");
      assert.ok(resp.url);
      mediaID = resp.media_id;
    });
  });

  it("Media/Download returns bytes", async (t) => {
    await run(t, async () => {
      if (!mediaID) skip("MAGNOLIA_MEDIA_FILE");
      const data = await client.downloadMedia(mediaID);
      assert.ok(data.byteLength > 0);
    });
  });

  it("Media/Thumbnail returns bytes or 404 for non-image/video", async (t) => {
    await run(t, async () => {
      if (!mediaID) skip("MAGNOLIA_MEDIA_FILE");
      try {
        const data = await client.getThumbnail(mediaID);
        assert.ok(data.byteLength > 0);
      } catch (err) {
        if (err instanceof APIError && err.statusCode === 404) return;
        throw err;
      }
    });
  });

  it("Media/Update sets description", async (t) => {
    await run(t, async () => {
      if (!mediaID) skip("MAGNOLIA_MEDIA_FILE");
      await client.updateMedia(mediaID, { description: `${PREFIX} media` });
    });
  });

  it("Media/BatchDelete removes a second upload", async (t) => {
    await run(t, async () => {
      requireEnv("MAGNOLIA_MEDIA_FILE");
      const extra = await client.uploadMediaFile(env("MAGNOLIA_MEDIA_FILE"));
      const result = await client.batchDeleteMedia([extra.media_id]);
      assert.equal(result.success_count, 1);
    });
  });

  it("Media/UploadChunked stores a file via chunked protocol", async (t) => {
    await run(t, async () => {
      requireEnv("MAGNOLIA_MEDIA_FILE");
      const filePath = env("MAGNOLIA_MEDIA_FILE");
      // 64 KiB chunks - forces multiple chunks even on small files
      const resp = await client.uploadMediaChunked(filePath, mediaTypeFor(filePath), 64 * 1024);
      assert.ok(resp.media_id, "empty media_id");
      await client.deleteMedia(resp.media_id);
    });
  });

  //  Conversations + Messages 

  it("Conversations/Create opens a direct conversation", async (t) => {
    await run(t, async () => {
      requireEnv("MAGNOLIA_TARGET_USER_ID");
      const resp = await client.createConversation("direct", [env("MAGNOLIA_TARGET_USER_ID")]);
      assert.ok(resp.conversation_id);
      convID = resp.conversation_id;
    });
  });

  it("Conversations/Get returns members list", async (t) => {
    await run(t, async () => {
      if (!convID) skip("MAGNOLIA_TARGET_USER_ID");
      const resp = await client.getConversation(convID);
      assert.equal(resp.conversation_id, convID);
      assert.ok(resp.members.some(m => m.user_id === env("MAGNOLIA_TARGET_USER_ID")));
    });
  });

  it("Conversations/List includes created conversation", async (t) => {
    await run(t, async () => {
      if (!convID) skip("MAGNOLIA_TARGET_USER_ID");
      const resp = await client.listConversations({ limit: 50 });
      assert.ok(resp.conversations.some(c => c.conversation_id === convID));
    });
  });

  it("Messages/Send delivers a message", async (t) => {
    await run(t, async () => {
      if (!convID) skip("MAGNOLIA_TARGET_USER_ID");
      const resp = await client.sendMessage(convID, `${PREFIX}_payload`);
      assert.ok(resp.message_id);
      msgID = resp.message_id;
    });
  });

  it("Messages/List includes the sent message", async (t) => {
    await run(t, async () => {
      if (!convID || !msgID) skip("MAGNOLIA_TARGET_USER_ID");
      const resp = await client.listMessages(convID);
      assert.ok(resp.messages.some(m => m.message_id === msgID));
    });
  });

  it("Messages/UnreadCounts returns counts map", async () => {
    const resp = await client.getUnreadCounts();
    assert.ok(typeof resp.counts === "object");
  });

  //  Proxy management (user session)

  it("Proxy/GetMyProxy returns proxy or skips if none paired", async (t) => {
    await run(t, async () => {
      try {
        const resp = await client.getMyProxy();
        assert.ok(resp.proxy_id, "empty proxy_id");
      } catch (err) {
        if (err instanceof APIError && err.statusCode === 404) {
          skip("no proxy account paired to this user");
        }
        throw err;
      }
    });
  });

  it("Proxy/ListPublicProxies returns an array", async () => {
    const resp = await client.listPublicProxies();
    assert.ok(Array.isArray(resp));
  });

  //  Messaging preferences

  it("MessagingPrefs/Toggle round-trips and restores", async () => {
    const prefs = await client.getMessagingPreferences();
    const original = prefs.accept_messages;
    try {
      const toggled = await client.updateMessagingPreferences(!original);
      assert.equal(toggled.accept_messages, !original);
    } finally {
      await client.updateMessagingPreferences(original);
    }
  });
});

//  Proxy session client tests

describe("Proxy session client", () => {
  it("ProxySession/Login-Me-Logout round-trip", async (t) => {
    await run(t, async () => {
      requireEnv("MAGNOLIA_BASE_URL", "MAGNOLIA_PROXY_USERNAME", "MAGNOLIA_PROXY_PASSWORD");
      const proxy = new MagnoliaProxySessionClient(env("MAGNOLIA_BASE_URL"), { timeoutMs: TIMEOUT_MS });
      const resp = await proxy.login(env("MAGNOLIA_PROXY_USERNAME"), env("MAGNOLIA_PROXY_PASSWORD"));
      assert.ok(resp.proxy_id, "empty proxy_id after login");
      try {
        const me = await proxy.me();
        assert.equal(me.proxy_id, resp.proxy_id);
        assert.equal(me.username, resp.username);
      } finally {
        await proxy.logout();
      }
    });
  });
});

//  HMAC proxy client tests

describe("HMAC proxy client", () => {
  let hmac: MagnoliaHMACClient;

  before(async (t) => {
    if (!env("MAGNOLIA_BASE_URL") || !env("MAGNOLIA_PROXY_ID") || !env("MAGNOLIA_HMAC_KEY")) {
      t.skip("MAGNOLIA_PROXY_ID / MAGNOLIA_HMAC_KEY not set");
      return;
    }
    hmac = new MagnoliaHMACClient(
      env("MAGNOLIA_BASE_URL"),
      env("MAGNOLIA_PROXY_ID"),
      env("MAGNOLIA_HMAC_KEY"),
      { timeoutMs: TIMEOUT_MS },
    );
  });

  it("HMAC/GetOrCreateConversation returns a conversation_id", async (t) => {
    await run(t, async () => {
      requireEnv("MAGNOLIA_TARGET_USER_ID");
      const resp = await hmac.getOrCreateConversation({ targetUserID: env("MAGNOLIA_TARGET_USER_ID") });
      assert.ok(resp.conversation_id);
    });
  });

  it("HMAC/SendMessage delivers a message", async (t) => {
    await run(t, async () => {
      requireEnv("MAGNOLIA_TARGET_USER_ID");
      const conv = await hmac.getOrCreateConversation({ targetUserID: env("MAGNOLIA_TARGET_USER_ID") });
      const resp = await hmac.sendMessage(conv.conversation_id, `${PREFIX}_hmac_payload`);
      assert.ok(resp.message_id);
    });
  });

  it("HMAC/CreatePost creates a draft post", async (t) => {
    await run(t, async () => {
      const contents: PostContentRequest[] = [
        { content_type: "text", display_order: 0, content: `${PREFIX} hmac post` },
      ];
      const resp = await hmac.createPost(contents, false, [PREFIX]);
      assert.ok(resp.post_id);
    });
  });

  it("HMAC/UploadMedia stores a file", async (t) => {
    await run(t, async () => {
      requireEnv("MAGNOLIA_MEDIA_FILE");
      const resp = await hmac.uploadMediaFile(env("MAGNOLIA_MEDIA_FILE"));
      assert.ok(resp.media_id);
      assert.ok(resp.url);
    });
  });
});

File: client/go/magnolia_test.go
// Integration tests for Client and HMACClient.
//
// Run against a live server:
//
//	MAGNOLIA_BASE_URL=https://... MAGNOLIA_USERNAME=alice MAGNOLIA_PASSWORD=... \
//	    go test -v -run TestIntegration ./...
//
// Required env vars:
//
//	MAGNOLIA_BASE_URL       server root, e.g. https://magnolia.example.com
//	MAGNOLIA_USERNAME       login identifier (username or email)
//	MAGNOLIA_PASSWORD       login password
//
// Optional env vars (related subtests are skipped when absent):
//
//	MAGNOLIA_PROXY_ID        )  both required for HMAC subtests
//	MAGNOLIA_HMAC_KEY        )

//	MAGNOLIA_PROXY_USERNAME  )  both required for proxy session subtests
//	MAGNOLIA_PROXY_PASSWORD  )

//	MAGNOLIA_MEDIA_FILE      path to any file; enables media upload/download subtests
//	MAGNOLIA_TARGET_USER_ID  enables conversation and message subtests
package magnolia_test

import (
	"errors"
	"fmt"
	"mime"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	magnolia "magnolia/client"
)

const testTimeout = 10 * time.Second

// mustEnv skips the test if the env var is unset and returns its value.
func mustEnv(t *testing.T, key string) string {
	t.Helper()
	if v := os.Getenv(key); v != "" {
		return v
	}
	t.Skipf("env var not set: %s", key)
	return ""
}

// optEnv returns the env var value or "".
func optEnv(key string) string { return os.Getenv(key) }

// check fatals the test if err is non-nil.
func check(t *testing.T, label string, err error) {
	t.Helper()
	if err != nil {
		t.Fatalf("%s: %v", label, err)
	}
}

// mediaTypeFor guesses the Magnolia media_type string from a file path.
func mediaTypeFor(path string) string {
	mt := mime.TypeByExtension(strings.ToLower(filepath.Ext(path)))
	switch {
	case strings.HasPrefix(mt, "image/"):
		return "image"
	case strings.HasPrefix(mt, "video/"):
		return "video"
	default:
		return "file"
	}
}

func TestIntegration(t *testing.T) {
	baseURL := mustEnv(t, "MAGNOLIA_BASE_URL")
	username := mustEnv(t, "MAGNOLIA_USERNAME")
	password := mustEnv(t, "MAGNOLIA_PASSWORD")

	c := magnolia.NewClient(baseURL, magnolia.WithTimeout(testTimeout))
	prefix := fmt.Sprintf("mgtest_%d", time.Now().Unix())

	// IDs accumulated across subtests; cleaned up in a single deferred block.
	var (
		userID    string
		postID    string
		commentID string
		mediaID   string
		convID    string
		msgID     string
	)

	t.Cleanup(func() {
		// Delete in dependency order so FK constraints are satisfied.
		if msgID != "" {
			_ = c.DeleteMessage(msgID)
		}
		if commentID != "" {
			_ = c.DeleteComment(commentID)
		}
		if mediaID != "" {
			_ = c.DeleteMedia(mediaID)
		}
		if convID != "" {
			_ = c.DeleteConversation(convID)
		}
		if postID != "" {
			_ = c.DeletePost(postID)
		}
		_ = c.Logout()
	})

	//  Auth 

	t.Run("Auth/Login", func(t *testing.T) {
		resp, err := c.Login(username, password)
		check(t, "Login", err)
		if resp.User.UserID == "" {
			t.Fatal("empty user_id in login response")
		}
		userID = resp.User.UserID
	})

	t.Run("Auth/Me", func(t *testing.T) {
		if userID == "" {
			t.Skip("login did not complete")
		}
		resp, err := c.Me()
		check(t, "Me", err)
		if resp.User.UserID != userID {
			t.Fatalf("Me user_id %q != login user_id %q", resp.User.UserID, userID)
		}
	})

	//  Profile 

	t.Run("Profile/Get", func(t *testing.T) {
		if userID == "" {
			t.Skip()
		}
		resp, err := c.GetProfile(userID)
		check(t, "GetProfile", err)
		if resp.UserID != userID {
			t.Fatalf("got user_id %q, want %q", resp.UserID, userID)
		}
	})

	t.Run("Profile/Update", func(t *testing.T) {
		if userID == "" {
			t.Skip()
		}
		orig, err := c.GetProfile(userID)
		check(t, "GetProfile", err)

		newBio := prefix + "_bio"
		updated, err := c.UpdateProfile(magnolia.UpdateProfileParams{Bio: &newBio})
		check(t, "UpdateProfile", err)
		if updated.Bio == nil || *updated.Bio != newBio {
			t.Fatalf("bio not updated, got %v", updated.Bio)
		}
		// Restore original bio regardless of test outcome.
		_, _ = c.UpdateProfile(magnolia.UpdateProfileParams{Bio: orig.Bio})
	})

	//  Posts 

	t.Run("Posts/Create", func(t *testing.T) {
		if userID == "" {
			t.Skip()
		}
		resp, err := c.CreatePost([]magnolia.PostContentRequest{
			{ContentType: "text", DisplayOrder: 0, Content: prefix + " post"},
		}, false, []string{prefix})
		check(t, "CreatePost", err)
		if resp.PostID == "" {
			t.Fatal("empty post_id")
		}
		postID = resp.PostID
	})

	t.Run("Posts/Get", func(t *testing.T) {
		if postID == "" {
			t.Skip()
		}
		resp, err := c.GetPost(postID)
		check(t, "GetPost", err)
		if resp.PostID != postID {
			t.Fatalf("got %q, want %q", resp.PostID, postID)
		}
	})

	t.Run("Posts/Update", func(t *testing.T) {
		if postID == "" {
			t.Skip()
		}
		newContent := prefix + " updated"
		resp, err := c.UpdatePost(postID, []magnolia.PostContentRequest{
			{ContentType: "text", DisplayOrder: 0, Content: newContent},
		}, nil, nil)
		check(t, "UpdatePost", err)
		if len(resp.Contents) == 0 || resp.Contents[0].Content != newContent {
			t.Fatalf("content not updated: %+v", resp.Contents)
		}
	})

	t.Run("Posts/Publish", func(t *testing.T) {
		if postID == "" {
			t.Skip()
		}
		resp, err := c.PublishPost(postID)
		check(t, "PublishPost", err)
		if !resp.IsPublished {
			t.Fatal("expected is_published=true after toggle")
		}
	})

	t.Run("Posts/List", func(t *testing.T) {
		if userID == "" {
			t.Skip()
		}
		resp, err := c.ListPosts(magnolia.ListPostsParams{Limit: 5})
		check(t, "ListPosts", err)
		if resp.Posts == nil {
			t.Fatal("nil posts slice")
		}
	})

	t.Run("Posts/Search", func(t *testing.T) {
		if postID == "" {
			t.Skip()
		}
		resp, err := c.SearchPosts(magnolia.SearchPostsParams{Q: prefix})
		check(t, "SearchPosts", err)
		if resp.Posts == nil {
			t.Fatal("nil posts slice")
		}
	})

	//  Comments 

	t.Run("Comments/Create", func(t *testing.T) {
		if postID == "" {
			t.Skip()
		}
		resp, err := c.CreateComment(postID, prefix+" comment", "", "", "", "")
		check(t, "CreateComment", err)
		if resp.CommentID == "" {
			t.Fatal("empty comment_id")
		}
		commentID = resp.CommentID
	})

	t.Run("Comments/Update", func(t *testing.T) {
		if commentID == "" {
			t.Skip()
		}
		edited := prefix + " edited"
		resp, err := c.UpdateComment(commentID, edited)
		check(t, "UpdateComment", err)
		if resp.Content != edited {
			t.Fatalf("got %q, want %q", resp.Content, edited)
		}
	})

	t.Run("Comments/List", func(t *testing.T) {
		if postID == "" || commentID == "" {
			t.Skip()
		}
		resp, err := c.ListComments(postID, magnolia.ListCommentsParams{})
		check(t, "ListComments", err)
		for _, cm := range resp.Comments {
			if cm.CommentID == commentID {
				return
			}
		}
		t.Fatalf("comment %q not found in listing", commentID)
	})

	//  Media 

	mediaFile := optEnv("MAGNOLIA_MEDIA_FILE")

	t.Run("Media/Upload", func(t *testing.T) {
		if mediaFile == "" {
			t.Skip("MAGNOLIA_MEDIA_FILE not set")
		}
		resp, err := c.UploadMedia(mediaFile, magnolia.UploadMediaParams{})
		check(t, "UploadMedia", err)
		if resp.MediaID == "" {
			t.Fatal("empty media_id")
		}
		mediaID = resp.MediaID
	})

	t.Run("Media/Download", func(t *testing.T) {
		if mediaID == "" {
			t.Skip()
		}
		data, err := c.DownloadMedia(mediaID)
		check(t, "DownloadMedia", err)
		if len(data) == 0 {
			t.Fatal("empty download response")
		}
	})

	t.Run("Media/Thumbnail", func(t *testing.T) {
		if mediaID == "" {
			t.Skip()
		}
		_, err := c.GetThumbnail(mediaID)
		if err != nil {
			var apiErr *magnolia.APIError
			if errors.As(err, &apiErr) && apiErr.StatusCode == 404 {
				t.Log("no thumbnail for this media type - acceptable")
				return
			}
			t.Fatalf("GetThumbnail: %v", err)
		}
	})

	t.Run("Media/Update", func(t *testing.T) {
		if mediaID == "" {
			t.Skip()
		}
		desc := prefix + " media"
		_, err := c.UpdateMedia(mediaID, &desc, nil)
		check(t, "UpdateMedia", err)
	})

	t.Run("Media/BatchDelete", func(t *testing.T) {
		if mediaFile == "" {
			t.Skip("MAGNOLIA_MEDIA_FILE not set")
		}
		// Upload a throwaway file specifically to batch-delete it.
		resp, err := c.UploadMedia(mediaFile, magnolia.UploadMediaParams{})
		check(t, "UploadMedia (batch target)", err)
		result, err := c.BatchDeleteMedia([]string{resp.MediaID})
		check(t, "BatchDeleteMedia", err)
		if result.SuccessCount != 1 {
			t.Fatalf("expected success_count=1, got %d", result.SuccessCount)
		}
	})

	t.Run("Media/UploadChunked", func(t *testing.T) {
		if mediaFile == "" {
			t.Skip("MAGNOLIA_MEDIA_FILE not set")
		}
		// 64 KiB chunks so multiple chunks are sent even for small files.
		resp, err := c.UploadMediaChunked(mediaFile, mediaTypeFor(mediaFile), 64*1024)
		check(t, "UploadMediaChunked", err)
		if resp.MediaID == "" {
			t.Fatal("empty media_id")
		}
		// Clean up immediately; not tracked in the outer deferred block.
		t.Cleanup(func() { _ = c.DeleteMedia(resp.MediaID) })
	})

	//  Conversations + Messages 

	targetUserID := optEnv("MAGNOLIA_TARGET_USER_ID")

	t.Run("Conversations/Create", func(t *testing.T) {
		if targetUserID == "" {
			t.Skip("MAGNOLIA_TARGET_USER_ID not set")
		}
		resp, err := c.CreateConversation("direct", []string{targetUserID}, "")
		check(t, "CreateConversation", err)
		if resp.ConversationID == "" {
			t.Fatal("empty conversation_id")
		}
		convID = resp.ConversationID
	})

	t.Run("Conversations/Get", func(t *testing.T) {
		if convID == "" {
			t.Skip()
		}
		resp, err := c.GetConversation(convID)
		check(t, "GetConversation", err)
		if resp.ConversationID != convID {
			t.Fatalf("got %q, want %q", resp.ConversationID, convID)
		}
		found := false
		for _, m := range resp.Members {
			if m.UserID == targetUserID {
				found = true
				break
			}
		}
		if !found {
			t.Fatalf("target user %q not in members list", targetUserID)
		}
	})

	t.Run("Conversations/List", func(t *testing.T) {
		resp, err := c.ListConversations(magnolia.ListConversationsParams{Limit: 5})
		check(t, "ListConversations", err)
		if resp.Conversations == nil {
			t.Fatal("nil conversations slice")
		}
	})

	t.Run("Messages/Send", func(t *testing.T) {
		if convID == "" {
			t.Skip()
		}
		resp, err := c.SendMessage(convID, prefix+"_payload", nil)
		check(t, "SendMessage", err)
		if resp.MessageID == "" {
			t.Fatal("empty message_id")
		}
		msgID = resp.MessageID
	})

	t.Run("Messages/List", func(t *testing.T) {
		if convID == "" || msgID == "" {
			t.Skip()
		}
		resp, err := c.ListMessages(convID, magnolia.ListMessagesParams{})
		check(t, "ListMessages", err)
		for _, m := range resp.Messages {
			if m.MessageID == msgID {
				return
			}
		}
		t.Fatalf("message %q not found in listing", msgID)
	})

	t.Run("Messages/UnreadCounts", func(t *testing.T) {
		resp, err := c.GetUnreadCounts()
		check(t, "GetUnreadCounts", err)
		if resp.Counts == nil {
			t.Fatal("nil counts map")
		}
	})

	//  Messaging preferences 

	t.Run("MessagingPrefs/GetAndToggle", func(t *testing.T) {
		if userID == "" {
			t.Skip()
		}
		prefs, err := c.GetMessagingPreferences()
		check(t, "GetMessagingPreferences", err)
		original := prefs.AcceptMessages

		toggled, err := c.UpdateMessagingPreferences(!original)
		check(t, "UpdateMessagingPreferences", err)
		if toggled.AcceptMessages == original {
			t.Fatal("preference was not toggled")
		}
		// Restore.
		_, _ = c.UpdateMessagingPreferences(original)
	})

	//  Proxy management (user session)

	t.Run("Proxy/GetMyProxy", func(t *testing.T) {
		if userID == "" {
			t.Skip()
		}
		resp, err := c.GetMyProxy()
		if err != nil {
			var apiErr *magnolia.APIError
			if errors.As(err, &apiErr) && apiErr.StatusCode == 404 {
				t.Skip("no proxy account paired to this user")
			}
			t.Fatalf("GetMyProxy: %v", err)
		}
		if resp.ProxyID == "" {
			t.Fatal("empty proxy_id")
		}
	})

	t.Run("Proxy/ListPublicProxies", func(t *testing.T) {
		if userID == "" {
			t.Skip()
		}
		proxies, err := c.ListPublicProxies()
		check(t, "ListPublicProxies", err)
		_ = proxies // may be empty; just verify no error
	})

	//  Proxy session

	proxyUsername := optEnv("MAGNOLIA_PROXY_USERNAME")
	proxyPassword := optEnv("MAGNOLIA_PROXY_PASSWORD")

	t.Run("ProxySession/LoginMeLogout", func(t *testing.T) {
		if proxyUsername == "" || proxyPassword == "" {
			t.Skip("MAGNOLIA_PROXY_USERNAME / MAGNOLIA_PROXY_PASSWORD not set")
		}
		p := magnolia.NewProxySessionClient(baseURL, magnolia.WithTimeout(testTimeout))
		loginResp, err := p.Login(proxyUsername, proxyPassword)
		check(t, "Login", err)
		if loginResp.ProxyID == "" {
			t.Fatal("empty proxy_id after login")
		}
		me, err := p.Me()
		check(t, "Me", err)
		if me.ProxyID != loginResp.ProxyID {
			t.Fatalf("Me proxy_id %q != login proxy_id %q", me.ProxyID, loginResp.ProxyID)
		}
		check(t, "Logout", p.Logout())
	})

	//  HMAC

	proxyID := optEnv("MAGNOLIA_PROXY_ID")
	hmacKey := optEnv("MAGNOLIA_HMAC_KEY")

	if proxyID == "" || hmacKey == "" {
		t.Log("MAGNOLIA_PROXY_ID / MAGNOLIA_HMAC_KEY not set - skipping HMAC subtests")
		return
	}

	h := magnolia.NewHMACClient(baseURL, proxyID, hmacKey, magnolia.WithTimeout(testTimeout))

	t.Run("HMAC/GetOrCreateConversation", func(t *testing.T) {
		if targetUserID == "" {
			t.Skip("MAGNOLIA_TARGET_USER_ID not set")
		}
		resp, err := h.GetOrCreateConversation(targetUserID, "")
		check(t, "GetOrCreateConversation", err)
		if resp.ConversationID == "" {
			t.Fatal("empty conversation_id")
		}
	})

	t.Run("HMAC/SendMessage", func(t *testing.T) {
		if targetUserID == "" {
			t.Skip("MAGNOLIA_TARGET_USER_ID not set")
		}
		conv, err := h.GetOrCreateConversation(targetUserID, "")
		check(t, "GetOrCreateConversation", err)
		resp, err := h.SendMessage(conv.ConversationID, prefix+"_hmac_payload", nil)
		check(t, "SendMessage", err)
		if resp.MessageID == "" {
			t.Fatal("empty message_id")
		}
	})

	t.Run("HMAC/CreatePost", func(t *testing.T) {
		resp, err := h.CreatePost([]magnolia.PostContentRequest{
			{ContentType: "text", DisplayOrder: 0, Content: prefix + " hmac post"},
		}, false, []string{prefix})
		check(t, "CreatePost", err)
		if resp.PostID == "" {
			t.Fatal("empty post_id")
		}
	})

	t.Run("HMAC/UploadMedia", func(t *testing.T) {
		if mediaFile == "" {
			t.Skip("MAGNOLIA_MEDIA_FILE not set")
		}
		resp, err := h.UploadMedia(mediaFile)
		check(t, "UploadMedia", err)
		if resp.MediaID == "" {
			t.Fatal("empty media_id")
		}
	})
}
